qiskit/circuit/qpy_serialization.py

```diff
@@ -0,0 +1,753 @@
+    if custom_definition_header[0] > 0:
+        for _ in range(custom_definition_header[0]):
+            custom_definition_raw = file_obj.read(CUSTOM_DEFINITION_SIZE)
+            custom_definition = struct.unpack(CUSTOM_DEFINITION_PACK,
+                                              custom_definition_raw)
+            (name_size, type_str, num_qubits,
+             num_clbits, has_custom_definition, size) = custom_definition
+            name = file_obj.read(name_size).decode('utf8').rstrip('utf8')
+            type_str = type_str.decode('utf8')
+            definition_circuit = None
+            if has_custom_definition:
+                definition_buffer = io.BytesIO(file_obj.read(size))
+                definition_circuit = _read_circuit(definition_buffer)
+            custom_instructions[name] = (type_str, num_qubits, num_clbits,
+                                         definition_circuit)
+    return custom_instructions
+
+
+def _write_instruction(file_obj, instruction_tuple, custom_instructions):
+    gate_class_name = instruction_tuple[0].__class__.__name__
+    if ((not hasattr(library, gate_class_name) and
+         not hasattr(circuit_mod, gate_class_name) and
+         not hasattr(extensions, gate_class_name) and
+         not hasattr(quantum_initializer, gate_class_name)) or
+            gate_class_name == 'Gate' or gate_class_name == 'Instruction'):
+        if instruction_tuple[0].name not in custom_instructions:
+            custom_instructions[
+                instruction_tuple[0].name] = instruction_tuple[0]
+        gate_class_name = instruction_tuple[0].name
+
+    has_condition = False
+    condition_register = ''.encode('utf8')
+    condition_value = 0
+    if instruction_tuple[0].condition:
+        has_condition = True
+        condition_register = instruction_tuple[0].condition[0].name.encode(
+            'utf8')
+        condition_value = instruction_tuple[0].condition[1]
+
+    gate_class_name = gate_class_name.encode('utf8')
+    instruction_raw = struct.pack(INSTRUCTION_PACK, len(gate_class_name),
+                                  len(instruction_tuple[0].params),
+                                  instruction_tuple[0].num_qubits,
+                                  instruction_tuple[0].num_clbits,
+                                  has_condition, len(condition_register),
+                                  condition_value)
+    file_obj.write(instruction_raw)
+    file_obj.write(gate_class_name)
+    file_obj.write(condition_register)
+    # Encode instruciton args
+    for qbit in instruction_tuple[1]:
+        qreg_name = qbit.register.name.encode('utf8')
+        instruction_arg_raw = struct.pack(INSTRUCTION_ARG_PACK,
+                                          'q'.encode('utf8'),
+                                          qbit.index,
+                                          len(qreg_name))
+        file_obj.write(instruction_arg_raw)
+        file_obj.write(qreg_name)
```

Comment:
To avoid the deprecated `.index` and `.register` in terra >= 0.17, would it be possible for `INSTRUCTION_ARG` to only contain the `type` and `index` (of the bit in `QuantumCircuit.{qubits,clbits}`)? `REGISTER` would also need to be updated to include which bit indices it has as members (and their order in the register).

Reply:
Done in https://github.com/Qiskit/qiskit-terra/pull/5578/commits/7c95f6831fb89adcaf6204a4a7fb6623547f0bd2